File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.swift
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public var val: Int
 *     public var left: TreeNode?
 *     public var right: TreeNode?
 *     public init() { self.val = 0; self.left = nil; self.right = nil; }
 *     public init(_ val: Int) { self.val = val; self.left = nil; self.right = nil; }
 *     public init(_ val: Int, _ left: TreeNode?, _ right: TreeNode?) {
 *         self.val = val
 *         self.left = left
 *         self.right = right
 *     }
 * }
 */
 
class Solution {
    func sumNumbers(_ root: TreeNode?) -> Int {
        return searchChild(root, 0)
    }
    
    func searchChild(_ node: TreeNode?, _ sum: Int) -> Int {
        guard let node = node else { return 0 }
        
        let sum = sum * 10 + node.val
        if node.left == nil && node.right == nil {
            return sum
        }
        
        let left = searchChild(node.left, sum)
        let right = searchChild(node.right, sum)
        
        return left + right
    }
}

File: 0129-sum-root-to-leaf-numbers/0129-sum-root-to-leaf-numbers.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int sumNumbers(TreeNode* root) {
        return search_child(root, 0);
    }
    
    int search_child(TreeNode* node, int sum) {
        if (!node) {
            return 0;
        }
        
        int current_sum = sum * 10 + node->val;
        
        if (!node->left && !node->right) {
            return current_sum;
        }
        
        int left = search_child(node->left, current_sum);
        int right = search_child(node->right, current_sum);
        
        return left + right;
    }
};
